stickermanager.py

```diff
@@ -0,0 +1,276 @@
+    yes = 0.0
+    no = 0.0
+    for vote in current_vote['votes'].values():
+        if vote.weight > 0:
+            yes += vote.weight
+        else:
+            no -= vote.weight
+    return Scores(fancy_round(yes - no), fancy_round(yes), fancy_round(no))
+
+
+def fancy_round(val: Union[float, int]) -> Union[float, int]:
+    if isinstance(val, float) and val.is_integer():
+        return int(val)
+    return round(val, 2)
+
+
+async def init(bot: TelegramClient) -> None:
+    @bot.on(events.NewMessage(pattern='#addsticker(?: (.+))?', chats=ALLOWED_CHATS))
+    async def start_poll(event: Union[events.NewMessage.Event, Message]) -> None:
+        if not event.is_reply:
+            return
+        elif current_vote:
+            await event.reply('There\'s already an ongoing sticker poll.')
+            return
+
+        async with current_vote_lock:
+            await start_poll_int(event)
+
+    async def start_poll_int(event: Union[events.NewMessage.Event, Message]) -> None:
+        global current_vote
+
+        if current_vote:
+            await event.reply('There\'s already an ongoing sticker poll.')
+            return
+
+        emoji = event.pattern_match.group(1) or '\u2728'
+        try:
+            _, sender_name = WEIGHTS[event.sender_id]
+        except KeyError:
+            return
+        orig_evt: Message = await event.get_reply_message()
+        # TODO add support for animated stickers
+        if not orig_evt.photo and (not orig_evt.sticker or
+                                   orig_evt.sticker.mime_type == 'application/x-tgsticker'):
+            return
+
+        filename = os.path.join(os.path.dirname(__file__), f'stickermanager.{int(time())}.dat')
+        await orig_evt.download_media(filename)
+
+        asyncio.ensure_future(event.delete(), loop=bot.loop)
+        current_vote = {
+            'chat': event.chat_id,
+            'sender_id': event.sender_id,
+            'sender_name': sender_name,
+            'score': 0,
+            'emoji': emoji,
+            'votes': {},
+            'filepath': filename,
+            'pack_link': f'https://t.me/addstickers/{STICKER_PACK_SHORT_NAME}',
+            'animated': orig_evt.sticker and orig_evt.sticker.mime_type == 'application/x-tgsticker'
+        }
+        reply_evt: Message = await orig_evt.reply(
+            POLL_TEMPLATE.format_map(get_template_data()),
+            buttons=[Button.inline(UP, b'+'), Button.inline(DOWN, b'-')],
+            parse_mode='html')
+        current_vote['poll'] = reply_evt.id
+        asyncio.ensure_future(reply_evt.pin(notify=True), loop=bot.loop)
+
+    @bot.on(events.CallbackQuery(chats=ALLOWED_CHATS, data=lambda data: data in (b'+', b'-')))
+    async def vote_poll(event: events.CallbackQuery.Event) -> None:
+        if not current_vote or current_vote['poll'] != event.message_id:
+            await event.answer('That poll is closed.')
+            return
+
+        async with current_vote_lock:
```

Comment:
The use of this lock seems error prone. Perhaps it should be abstracted into its own function (e.g. `locked_vote_poll_int`)? Or better, make `vote_poll_int` private or even local. Or better, use a `@locked(current_vote_lock)` decorator.